@xen-orchestra/backups/RemoteAdapter.js

```diff
@@ -551,4 +557,273 @@ exports.RemoteAdapter = class RemoteAdapter {
+      const { mode } = metadata
+      if (mode === 'full') {
+        const linkedXva = resolve(vmDir, metadata.xva)
+
+        if (xvas.has(linkedXva)) {
+          unusedXvas.delete(linkedXva)
+        } else {
+          onLog(`the XVA linked to the metadata ${json} is missing`)
+          if (remove) {
+            await handler.unlink(json)
+          }
+        }
+      } else if (mode === 'delta') {
+        const linkedVhds = (() => {
+          const { vhds } = metadata
+          return Object.keys(vhds).map(key => resolve(vmDir, vhds[key]))
+        })()
+
+        // FIXME: find better approach by keeping as much of the backup as
+        // possible (existing disks) even if one disk is missing
+        if (linkedVhds.every(_ => vhds.has(_))) {
+          linkedVhds.forEach(_ => unusedVhds.delete(_))
+        } else {
+          onLog(`Some VHDs linked to the metadata ${json} are missing`)
+          if (remove) {
+            await handler.unlink(json)
+          }
+        }
+      }
+    })
+
+    // TODO: parallelize by vm/job/vdi
+    const unusedVhdsDeletion = []
+    {
+      // VHD chains (as list from child to ancestor) to merge indexed by last
+      // ancestor
+      const vhdChainsToMerge = { __proto__: null }
+
+      const toCheck = new Set(unusedVhds)
+
+      const getUsedChildChainOrDelete = vhd => {
+        if (vhd in vhdChainsToMerge) {
+          const chain = vhdChainsToMerge[vhd]
+          delete vhdChainsToMerge[vhd]
+          return chain
+        }
+
+        if (!unusedVhds.has(vhd)) {
+          return [vhd]
+        }
+
+        // no longer needs to be checked
+        toCheck.delete(vhd)
+
+        const child = vhdChildren[vhd]
+        if (child !== undefined) {
+          const chain = getUsedChildChainOrDelete(child)
+          if (chain !== undefined) {
+            chain.push(vhd)
+            return chain
+          }
+        }
+
+        onLog(`the VHD ${vhd} is unused`)
+        if (remove) {
+          unusedVhdsDeletion.push(handler.unlink(vhd))
+        }
+      }
+
+      toCheck.forEach(vhd => {
+        vhdChainsToMerge[vhd] = getUsedChildChainOrDelete(vhd)
+      })
+
+      Object.keys(vhdChainsToMerge).forEach(key => {
+        const chain = vhdChainsToMerge[key]
+        if (chain !== undefined) {
+          unusedVhdsDeletion.push(this._mergeVhdChain(chain, { onLog, remove, merge }))
+        }
+      })
+    }
+
+    await Promise.all([
+      unusedVhdsDeletion,
+      asyncMap(unusedXvas, path => {
+        onLog(`the XVA ${path} is unused`)
+        return remove && handler.unlink(path)
+      }),
+      asyncMap(xvaSums, path => {
+        // no need to handle checksums for XVAs deleted by the script, they will be handled by `unlink()`
+        if (!xvas.has(path.slice(0, -'.checksum'.length))) {
+          onLog(`the XVA checksum ${path} is unused`)
+          return remove && handler.unlink(path)
+        }
+      }),
+    ])
+  }
```

Comment:
This file is getting huge, I'd like to find a way to improve the situation.

What do you think of moving this into its own file?

```js
import { cleanVm } from './RemoteAdapter/_cleanVms.js'
RemoteAdapter.prototype.cleanVm = cleanVm
```
Remote